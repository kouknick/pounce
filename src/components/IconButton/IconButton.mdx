An Iconbutton can be solid and have different sizes:

```typescript jsx
<Box>
  <SimpleGrid columns={3} gap={4} mb={2}>
    <Box>
      <IconButton size="small" icon="user" aria-label="User" />
    </Box>
    <Box>
      <IconButton size="small" icon="user" variantColor="teal" aria-label="User" />
    </Box>
    <Box>
      <IconButton
        size="small"
        icon="user"
        variantColor="violet"
        variantBorderStyle="circle"
        aria-label="User"
      />
    </Box>
  </SimpleGrid>
  <SimpleGrid columns={3} gap={4} mb={2}>
    <Box>
      <IconButton size="medium" icon="user" aria-label="User" />
    </Box>
    <Box>
      <IconButton size="medium" icon="user" variantColor="teal" aria-label="User" />
    </Box>
    <Box>
      <IconButton
        size="medium"
        icon="user"
        variantColor="violet"
        variantBorderStyle="circle"
        aria-label="User"
      />
    </Box>
  </SimpleGrid>
  <SimpleGrid columns={3} gap={4}>
    <Box>
      <IconButton icon="user" aria-label="User" />
    </Box>
    <Box>
      <IconButton icon="user" variantColor="teal" aria-label="User" />
    </Box>
    <Box>
      <IconButton icon="user" variantColor="violet" variantBorderStyle="circle" aria-label="User" />
    </Box>
  </SimpleGrid>
</Box>
```

An IconButton can be `ghost`:

```typescript jsx
<SimpleGrid columns={4} gap={4}>
  <IconButton variant="ghost" icon="user" aria-label="User" variantColor="navyblue" />
  <IconButton
    variant="ghost"
    icon="user"
    aria-label="User"
    variantColor="teal"
    variantBorderStyle="circle"
  />
  <IconButton variant="ghost" icon="user" aria-label="User" variantColor="navyblue" active />
  <IconButton
    variant="ghost"
    icon="user"
    aria-label="User"
    variantColor="teal"
    variantBorderStyle="circle"
    active
  />
</SimpleGrid>
```

An IconButton can be `outline`:

```typescript jsx
<SimpleGrid columns={4} gap={4}>
  <IconButton variant="outline" icon="user" aria-label="User" variantColor="navyblue" />
  <IconButton
    variant="outline"
    icon="user"
    aria-label="User"
    variantColor="teal"
    variantBorderStyle="circle"
  />
  <IconButton variant="outline" icon="user" aria-label="User" variantColor="navyblue" active />
  <IconButton
    variant="outline"
    icon="user"
    aria-label="User"
    variantColor="teal"
    variantBorderStyle="circle"
    active
  />
</SimpleGrid>
```

An IconButton can have no styling:

```typescript jsx
<SimpleGrid columns={4} gap={4}>
  <IconButton variant="unstyled" icon="user" aria-label="User" />
  <IconButton variant="unstyled" icon="user" aria-label="User" variantBorderStyle="circle" />
</SimpleGrid>
```
