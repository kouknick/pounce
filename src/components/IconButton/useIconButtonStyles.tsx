import React from 'react';
import { addOpacity } from '../../utils/helpers';
import { AbstractButtonProps } from '../AbstractButton';
import { getSolidButtonStyles, getThemeColor } from '../Button/useButtonStyles';
import useTheme from '../../utils/useTheme';
import { IconButtonProps } from './IconButton';
import { Theme } from '../../theme';
import { ButtonProps } from '../Button';

type ButtonColorVariant = ButtonProps['variantColor'];
type UseIconButtonStyles = Required<Pick<IconButtonProps, 'variantColor' | 'variant' | 'size'>>;

export const getUnstyledButtonStyles = (theme: Theme) => {
  return {
    _focus: {
      borderRadius: 'circle' as const,
      backgroundColor: addOpacity(theme.colors.white, 0.1),
    },
  };
};

export const getGhostButtonStyles = (theme: Theme, variantColor: ButtonColorVariant) => {
  const themeColorKey = getThemeColor(variantColor);
  const themeColor = theme.colors[themeColorKey];

  return {
    transition: 'background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms',
    borderRadius: 'circle' as const,
    border: '1px solid',
    borderColor: 'transparent',
    backgroundColor: 'transparent',
    _hover: {
      backgroundColor: addOpacity(themeColor, 0.3),
    },
    _focus: {
      backgroundColor: addOpacity(themeColor, 0.3),
    },
    _active: {
      backgroundColor: themeColor,
    },
  };
};

const useIconButtonStyles = ({
  variantColor,
  variant,
  size,
}: UseIconButtonStyles): AbstractButtonProps => {
  const theme = useTheme();

  const styles = React.useMemo(() => {
    switch (variant) {
      case 'ghost':
        return getGhostButtonStyles(theme, variantColor);
      case 'unstyled':
        return getUnstyledButtonStyles(theme);
      case 'solid':
      default:
        return getSolidButtonStyles(theme, variantColor);
    }
  }, [variantColor, variant]);

  return {
    p: size === 'small' ? 1 : 3,
    outline: 'none',
    ...styles,
  } as AbstractButtonProps;
};

export default useIconButtonStyles;
