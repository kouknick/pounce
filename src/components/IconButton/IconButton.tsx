import React from 'react';
import Icon, { IconProps } from '../Icon';
import useIconButtonStyles from './useIconButtonStyles';
import { ReactAttributes } from '../Box';
import AbstractButton from '../AbstractButton';

export type IconButtonProps = ReactAttributes<React.ButtonHTMLAttributes<HTMLButtonElement>> & {
  /** The size of the icon button */
  size?: 'small' | 'large';

  /** The text associated with the icon button */
  'aria-label': string;

  /** The icon present on the button  */
  icon: IconProps['type'];

  /** The style of the icon button */
  variant?: 'solid' | 'ghost' | 'unstyled';

  /** The color scheme of the button for solid variants */
  variantColor?:
    | 'blue'
    | 'violet'
    | 'teal'
    | 'green'
    | 'orange'
    | 'red'
    | 'gray'
    | 'darkgray'
    | 'navyblue';

  /** Whether the button is disabled */
  disabled?: boolean;

  /** Whether the button should have a loading spinner next to it */
  loading?: boolean;

  /** Whether the button should always be marked as active. Helpful for menus. */
  active?: boolean;
};

/** A wrapper that makes an <a href="/#/Icon">Icon<a> component be clickable */
export const IconButton = React.forwardRef<HTMLButtonElement, IconButtonProps>(function IconButton(
  { variantColor = 'blue', active = false, variant = 'solid', size = 'large', icon, ...rest },
  ref
) {
  const styles = useIconButtonStyles({ variantColor, variant, size });

  return (
    <AbstractButton
      aria-pressed={active !== undefined ? active : undefined}
      data-active={active || undefined}
      {...styles}
      {...rest}
      ref={ref}
    >
      <Icon type={icon} size="small" display="block" />
    </AbstractButton>
  );
});

export default IconButton;
