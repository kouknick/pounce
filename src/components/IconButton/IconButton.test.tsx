import React from 'react';
import { renderWithTheme } from 'test-utils';
import IconButton from './index';

describe('IconButton', () => {
  it('renders `small` size correctly', () => {
    const { container } = renderWithTheme(<IconButton size="small" icon="add" aria-label="Test" />);

    expect(container).toMatchSnapshot();
  });

  it('renders `medium` size correctly', () => {
    const { container } = renderWithTheme(
      <IconButton size="medium" icon="add" aria-label="Test" />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders `large` size correctly', () => {
    const { container } = renderWithTheme(<IconButton size="large" icon="add" aria-label="Test" />);

    expect(container).toMatchSnapshot();
  });

  it('renders `variantColor`  and `variants` correctly', () => {
    const { container } = renderWithTheme(
      <>
        <IconButton variantColor="teal" variant="solid" size="large" icon="add" aria-label="Test" />
        <IconButton variantColor="red" variant="ghost" icon="brackets" aria-label="Test red" />
        <IconButton variantColor="darkblue" icon="timer" aria-label="Test timer" />
      </>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders `square` variantBorderStyle correctly', () => {
    const { container } = renderWithTheme(
      <IconButton variantBorderStyle="square" icon="add" aria-label="Test" />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders `circle` variantBorderStyle correctly', () => {
    const { container } = renderWithTheme(
      <IconButton variantBorderStyle="circle" icon="add" aria-label="Test" />
    );

    expect(container).toMatchSnapshot();
  });
});
