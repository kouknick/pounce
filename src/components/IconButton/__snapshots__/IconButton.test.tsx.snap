// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton renders \`large\` size correctly 1`] = `
.pounce-1 {
  outline: none;
  line-height: 1;
  height: 46px;
  width: 46px;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.pounce-1:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.pounce-1:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.pounce-1:active,
.pounce-1[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

.pounce-0 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  color: currentColor;
}

<div>
  <button
    aria-label="Test"
    aria-pressed="false"
    class="pounce-1"
    type="button"
  >
    <svg
      class="pounce-0"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 13.1429H13.1429V20H10.8571V13.1429H4V10.8571H10.8571V4H13.1429V10.8571H20V13.1429Z"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton renders \`medium\` size correctly 1`] = `
.pounce-1 {
  outline: none;
  line-height: 1;
  height: 32px;
  width: 32px;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.pounce-1:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.pounce-1:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.pounce-1:active,
.pounce-1[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

.pounce-0 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: currentColor;
}

<div>
  <button
    aria-label="Test"
    aria-pressed="false"
    class="pounce-1"
    type="button"
  >
    <svg
      class="pounce-0"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 13.1429H13.1429V20H10.8571V13.1429H4V10.8571H10.8571V4H13.1429V10.8571H20V13.1429Z"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton renders \`small\` size correctly 1`] = `
.pounce-1 {
  outline: none;
  line-height: 1;
  height: 24px;
  width: 24px;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.pounce-1:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.pounce-1:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.pounce-1:active,
.pounce-1[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

.pounce-0 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  color: currentColor;
}

<div>
  <button
    aria-label="Test"
    aria-pressed="false"
    class="pounce-1"
    type="button"
  >
    <svg
      class="pounce-0"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 13.1429H13.1429V20H10.8571V13.1429H4V10.8571H10.8571V4H13.1429V10.8571H20V13.1429Z"
      />
    </svg>
  </button>
</div>
`;
